enterprise_access/apps/content_assignments/api.py

```diff
@@ -294,6 +295,24 @@ def _get_content_title(assignment_configuration, content_key):
     return content_metadata.get('title')
 
 
+def _try_populate_assignment_lms_user_id(assignment):
+    """
+    Try to populate the assignment.lms_user_id field based on a matching User.
+
+    Notes:
+    * This function does NOT save() the assignment record.
+    * This is a best-effort only; most of the time a User will not exist, and this function is a no-op.
+    * If multiple User records match based on email, choice is non-deterministic.
+
+    Args:
+        LearnerContentAssignment: the assignment on which to update the lms_user_id field.
+    """
+    if not assignment.lms_user_id:
+        first_user = User.objects.filter(email=assignment.learner_email).first()
```

Comment:
My only concern with this is that we'll end up doing one select against the `User` table for each provided learner email.  Can we prefetch the users in bulk, store them in a dictionary of `{email: User record}`, and use that dictionary to populate the `lms_user_id`?

Reply:
I was trying to brainstorm how to do this, but what comes to mind aren't ideal (but still better than what's currently written).
1. Prefetch only the users relevant to the requested emails.  I think this would be sensitive to limitations in how long one line of SQL can be, since the statement length is a function of the request size.
2. Prefetch all users in the system to build a mapping including users we don't care about.  I think this would result in a full table scan on a potentially very large table (User).
3. \# 1, but prefetch users in batches.  This would work but the code complexity seems too heavy for such a simple task.
4. Bulk create assignments without lms_user_id, then followup fast with a fancy ORM query to do an update+join (second write) to populate the lms_user_id on the newly created assignments.  The downside here is two writes for an API endpoint which is supposed to be super reliable and atomic.

I'm leaning towards 4 but LMK what you think about introducing a second write to the transaction.

Reply:
we discussed during standup and landed on option 3 because we have some prior art doing it this way in license-manager already.